### ISSUE TITLE
[BUG] RecursionError when running Pylint in parallel, unpickle setoptconf

### ISSUE THREAD

**Issue by rpatterson** [PyCQA/prospector#573]
**Describe the bug**

When Pylint runs parallel checks, it pickles the linter object to pass to it's workers which in turn unpickle it.  Under Prospector, the found files object includes filters which are closures which include the Prospector config.  The `setoptconf` config object seems to get unpickled without it's `_settings` instance attribute which results in endless recursion.

**To Reproduce**

```ShellSession
$ git clone -b docker https://github.com/rpatterson/python-project-structure.git
$ cd python-project-structure
$ make .env
$ editor .env  # Add the DOCKER_GID from your /etc/group
$ rm ./.prospector.yaml
$ docker compose run --rm docker bash
$ make build-docker-3.11
$ docker compose run --rm python-project-structure-devel tox exec --no-recreate-pkg --skip-pkg-install -e py311 -- prospector -X --direct-tool-stdout ./src/pythonprojectstructure/
...
Exception in initializer:
Traceback (most recent call last):
  File "/usr/local/lib/python3.11/concurrent/futures/process.py", line 235, in _process_worker
    initializer(*initargs)
  File "/usr/local/src/python-project-structure/.tox/py311/lib/python3.11/site-packages/pylint/lint/parallel.py", line 48, in _worker_initialize
    _worker_linter = dill.loads(linter)
		     ^^^^^^^^^^^^^^^^^^
  File "/usr/local/src/python-project-structure/.tox/py311/lib/python3.11/site-packages/dill/_dill.py", line 286, in loads
    return load(file, ignore, **kwds)
	   ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/src/python-project-structure/.tox/py311/lib/python3.11/site-packages/dill/_dill.py", line 272, in load
    return Unpickler(file, ignore=ignore, **kwds).load()
	   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/src/python-project-structure/.tox/py311/lib/python3.11/site-packages/dill/_dill.py", line 419, in load
    obj = StockUnpickler.load(self)
	  ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/src/python-project-structure/.tox/py311/lib/python3.11/site-packages/setoptconf/config.py", line 33, in __getattr__
    if name in self._settings:
	       ^^^^^^^^^^^^^^
  File "/usr/local/src/python-project-structure/.tox/py311/lib/python3.11/site-packages/setoptconf/config.py", line 33, in __getattr__
    if name in self._settings:
	       ^^^^^^^^^^^^^^
  File "/usr/local/src/python-project-structure/.tox/py311/lib/python3.11/site-packages/setoptconf/config.py", line 33, in __getattr__
    if name in self._settings:
	       ^^^^^^^^^^^^^^
  [Previous line repeated 970 more times]
RecursionError: maximum recursion depth exceeded
...
```

**Expected behavior**

Run successfully and report results.

**Environment:**

- OS: Alpine Linux v3.17 via the `docker` image (Pop!_OS 22.04 LTS is the host for the example above)
- Tool: pylint
- Prospector version: `prospector[with_everything]==1.8.3`
- Python version: 3.11.1

**Additional context**

See the `./requirements/py311/devel.txt` file in the repository cloned above for full dependencies and versions.

**Comment by rpatterson** [comment]
There's some variable here I can't isolate.  I can't reproduce the issue when I run the same `$ tox exec` directly on my laptop as opposed to in the `FROM python` container.  I can't reproduce the issue with the following `./docker-compose.yml`:

```
version: "3.8"
services:
  python:
    image: "python:3.11"
    working_dir: "/usr/local/src"
    command: >-
      bash -xeu -c '
      mkdir -pv "./grault/" &&
      echo "from . import corge" >"./grault/__init__.py" &&
      echo "from . import garply" >"./grault/corge.py" &&
      echo "import sys" >"./grault/garply.py" &&
      echo "[tox]" >"./tox.ini" &&
      echo "envlist = py311" >>"./tox.ini" &&
      echo "[testenv]" >>"./tox.ini" &&
      echo "deps = prospector[with_everything]" >>"./tox.ini" &&
      pip install "tox" &&
      tox exec --no-recreate-pkg --skip-pkg-install -- prospector -X --direct-tool-stdout'
```

Further, I don't understand why the `setoptconf` config object is unpickled without it's `_settings` instance attribute because that should be included in the instance state when pickled.  So I understand if you want to close this, but it also may be a genuine bug.  I at least wanted to get this traceback in a searchable place in case others encounter this.  If anyone can help me understand what the difference is here between when it works and doesn't, that would be great.

**Comment by rpatterson** [comment]
I also tried changing to `FROM ubuntu` as a base image, the resulting image reproduces the same error, so I don't think it's the `python` base image.

**Comment by rpatterson** [comment]
Finally!  I commented out the following in `./pyproject.toml`:

```
[tool.pylint.MASTER]
# jobs = 0
```

Then prospector succeeded in the container.  I discovered that I had a `~/.config/pylintrc` file containing defaults and when I moved that aside and *un-commented* the above I was able to reproduce the issue locally on my laptop.  So that explains the difference I was seeing.  IOW, you can reproduce this bug by setting `jobs = 0` in a configuration file used by pylint.  So I would consider this a bug after all.

**Comment by Pierre-Sassoulas** [comment]
Thank you for the analysis, it does seem like a bug (but maybe in pylint ?).

**Comment by rpatterson** [comment]
> Thank you for the analysis, it does seem like a bug (but maybe in pylint ?).

I don't think so, `$ pylint --jobs 0` ran fine in the same environments that reproduce the exeception.  See also my description of which objects are being unpickled that cause the endless recursion, those come *only* from Prospector.

**Comment by rpatterson** [comment]
```
>>> from setoptconf import config
>>> config.Configuration([]).__getstate__()
{'_parent': None, '_settings': {}}
```

**Comment by Polsaker** [comment]
Ran into this same error when setting `pylint.options.jobs` to any value in `.prospector.yaml`

**Comment by koniiiik** [comment]
After spending a couple of days trying to figure out what is going on here, I think I have a pretty decent understanding.

The root problem is that `setoptconf` is incompatible with pickle, but it is referenced from `ProspectorLinter`, which, in turn needs to be picklable by `dill` in order for parallel execution to work. `setoptconf` also doesn't have a whole lot of activity, so I have no idea how feasible it is to add pickling support there. That would probably be the most “correct” way to fix this.

The regression happened in 240e0ca1bdfce40f6a225e4b600805dfeb30c14c, when `ProspectorConfig.make_exclusion_filter` was introduced. `make_exclusion_filter` returns a closure that references `self`, and `self` has a reference to the `setoptconf` object. The filter closure is, in turn, referenced by `FileFinder`, which is assigned to `ProspectorLinter._files`. When `dill` pickles a function object, it also pickles the values it closes over, which includes `ProspectorConfig` in this case.

I'm currently not sure how exactly to fix this in the short term, but for the time being, breaking the reference to `ProspectorConfig` in the closure should alleviate the issue.

Continuing to investigate…

**Comment by carlio** [comment]
`setoptconf` hasn't been updated for a long time so I ended up making a fork here : https://github.com/carlio/setoptconf-tmp

The idea was that this was temporary to eventually replace it (I wanted to start using `click`) but so far my attempts just live on a dead branch :-)

One idea is that `make_exclusion_filter` doesn't actually need the reference to `self`, it only needs the `workdir` (which is a `Path` and I believe picklable?). 

The config once created is never changed after the invocation of prospector, therefore one solution could be to change this method to avoid the reference to `self` and therefore to `ProspectorConfig`.

Just an idea, I haven't investigated this though.

**Comment by koniiiik** [comment]
I have a patch that fixes the issue for me, and makes prospector run successfully on our code base where it would crash previously, but I ran into another issue. I wanted to create a test that runs pylint in parallel mode, so that issues like this are caught earlier, but I'm struggling to reproduce the crash in a clean test environment.

For what it's worth, just closing over the two specific attributes that `_filter` makes use of, instead of the entire `self`, is enough to avoid the pickling issue.

**Comment by rpatterson** [comment]
Running PyLint in parallel with `jobs = 0` is working for me since the changes in #577 were released.
